src/weights.jl

```diff
@@ -473,116 +473,50 @@ _mean(A::AbstractArray{T}, w::AbstractWeights{W}, dims::Nothing) where {T<:Numbe
-"""
-wmedian(v::RealVector, w::RealVector) = median(v, weights(w))
-wmedian(v::RealVector, w::AbstractWeights{<:Real}) = median(v, w)
 
 ###### Weighted quantile #####
-
-
 """
     quantile(v, w::AbstractWeights, p)
 
 Compute the weighted quantiles of a vector `v` at a specified set of probability
 values `p`, using weights given by a weight vector `w` (of type `AbstractWeights`).
 Weights must not be negative. The weights and data vectors must have the same length.
-
-With [`FrequencyWeights`](@ref), the function returns the same result as
-`quantile` for a vector with repeated values.
-With non `FrequencyWeights`,  denote ``N`` the length of the vector, ``w`` the vector of weights,
-``h = p (\\sum_{i<= N}w_i - w_1) + w_1`` the cumulative weight corresponding to the
-probability ``p`` and ``S_k = \\sum_{i<=k}w_i`` the cumulative weight for each
+`NaN` is returned if `x` contains any `NaN` values. An error is raised if `w` contains 
+any `NaN` values.
+
+With [`FrequencyWeights`](@ref), the function returns the same result as the unweighted 
+`quantile` applied to a vector with repeated values. The function returns an error if the 
+weights are not of type `AbstractVector{<:Integer}`.
+With non `FrequencyWeights`, denote ``N`` the length of the vector, ``w`` the vector of 
+weights, ``h = p (\\sum_{i<= N}w_i - w_1) + w_1`` the cumulative weight corresponding to 
+the probability ``p`` and ``S_k = \\sum_{i<=k}w_i`` the cumulative weight for each
 observation, define ``v_{k+1}`` the smallest element of `v` such that ``S_{k+1}``
-is strictly superior to ``h``. The weighted ``p`` quantile is given by ``v_k + \\gamma (v_{k+1} -v_k)``
-with  ``\\gamma = (h - S_k)/(S_{k+1}-S_k)``. In particular, when `w` is a vector
-of ones, the function returns the same result as `quantile`.
+is strictly superior to ``h``. The weighted ``p`` quantile is given by 
+``v_k + \\gamma (v_{k+1} -v_k)`` with  ``\\gamma = (h - S_k)/(S_{k+1}-S_k)``. 
 """
+
 function quantile(v::RealVector{V}, w::AbstractWeights{W}, p::RealVector) where {V,W<:Real}
     # checks
     isempty(v) && error("quantile of an empty array is undefined")
     isempty(p) && throw(ArgumentError("empty quantile array"))
 
     w.sum == 0 && error("weight vector cannot sum to zero")
-    length(v) == length(w) || error("data and weight vectors must be the same size, got $(length(v)) and $(length(w))")
+    length(v) == length(w) || error("data and weight vectors must be the same size, 
+        got $(length(v)) and $(length(w))")
     for x in w.values
         isnan(x) && error("weight vector cannot contain NaN entries")
         x < 0 && error("weight vector cannot contain negative entries")
     end
 
 
+    isa(w, FrequencyWeights) && !(eltype(w) <: Integer) && any(!isinteger, w) && 
+        error("The values of the vector of `FrequencyWeights` must be numerically equal to 
+        integers. Use `ProbabilityWeights` or `AnalyticWeights` instead.")
+
+    @inbounds for x in v
```

Comment:
To be 100% certain the return type is type-stable, better call `fill!(x, out)` below, after allocating `out`.